File: src/main/java/dsl/TypedSourceBuilder.java
package dsl;

public class TypedSourceBuilder<S> {
    private final Class<S> sourceClass;
    private Format sourceFormat = Format.JAVA_BEAN;

    TypedSourceBuilder(Class<S> sourceClass) {
        this.sourceClass = sourceClass;
    }

    public TypedSourceBuilder<S> withFormat(Format format) {
        this.sourceFormat = format;
        return this;
    }

    public <T> TypedTargetBuilder<S, T> to(Class<T> targetClass) {
        return new TypedTargetBuilder<>(sourceClass, sourceFormat, targetClass);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/typed/TypedSourceBuilder.kt
package xyz.mahmoudahmed.dsl.typed

import xyz.mahmoudahmed.format.Format

class TypedSourceBuilder<S>(
    private val sourceClass: Class<S>
) {
    private var sourceFormat: Format = Format.JSON

    fun withFormat(format: Format): TypedSourceBuilder<S> {
        this.sourceFormat = format
        return this
    }

    fun <T> to(targetClass: Class<T>): TypedTargetBuilder<S, T> {
        return TypedTargetBuilder(sourceClass, sourceFormat, targetClass)
    }
}